### PR TITLE
mathext: Implemented associate elliptic integrals B(m),D(m).

### PR DESCRIPTION
**Pull request by reconditematter**
Added an implementation of 2 associate complete elliptic integrals B(m),D(m). Added tests for the implemented integrals. B(m) and D(m) can be expressed via K(m) and E(m), but with cancellation problems. This implementation, just like my previous implementation of K(m) and E(m), based on the paper by Fukushima.
Reference [https://doi.org/10.1016/j.cam.2014.12.038](https://doi.org/10.1016/j.cam.2014.12.038)
Preprint of the above paper is attached.
[fukushima.pdf](https://github.com/gonum/gonum/files/1632910/fukushima.pdf)

### CODE REVIEW

File: mathext/ell_complete.go [hunk 1]
```diff
@@ -155,3 +155,168 @@ func CompleteE(m float64) float64 {
 
 	return 1
 }
+
+// CompleteB computes an associate complete elliptic integral of the 2nd kind, 0≤m≤1.
+// It returns math.NaN() if m is not in [0,1].
+//
```
Comment (btracey) [inline mathext/ell_complete.go]: Could you remove the newlines and change `\theta` to `θ`

File: mathext/ell_complete.go [hunk 2]
```diff
@@ -155,3 +155,168 @@ func CompleteE(m float64) float64 {
...
+//	B(m) = \int_{0}^{\pi/2} {\cos^2\theta} / {\sqrt{1-m{\sin^2\theta}}} d\theta
+//
+// See: http://dx.doi.org/10.1016/j.cam.2014.12.038 for the computation method.
```
Comment (btracey) [inline mathext/ell_complete.go]: Move implementation details inside the function. Could you enhance the citation? The linked article is 50 pages long. In particular, I don't see the numbers below in any of the tables.

File: mathext/ell_complete.go [hunk 3]
```diff
@@ -155,3 +155,168 @@ func CompleteE(m float64) float64 {
...
+		q := (26251.5678902584870 + t*(168.788023807915689+t*0.352150236262724288)) / (26065.7912239203873 + t*(353.916840382280456+t))
+		return p + q
+	}
+
+	return 1
+}
+
+// CompleteD computes an associate complete elliptic integral of the 2nd kind, 0≤m≤1.
+// It returns math.NaN() if m is not in [0,1].
+//
```
Comment (btracey) [inline mathext/ell_complete.go]: Same comments as above.

**Comment by btracey** [comment]
Thanks for a PR

File: mathext/ell_complete.go [hunk 4]
```diff
@@ -8,13 +8,17 @@ import (
...
-// See: http://dx.doi.org/10.1016/j.cam.2014.12.038 for the computation method.
+//  K(m) = \int_{0}^{π/2} 1/{\sqrt{1-m{\sin^2θ}}} dθ
 func CompleteK(m float64) float64 {
+	// Reference:
+	// Toshio Fukushima, Precise and fast computation of complete elliptic integrals
+	// by piecewise minimax rational function approximation,
+	// Journal of Computational and Applied Mathematics, Volume 282, 2015, Pages 71-76.
+	// https://doi.org/10.1016/j.cam.2014.12.038
+	// Original Fortran code available at:
+	// https://www.researchgate.net/publication/295857819_xceitxt_F90_package_of_complete_elliptic_integral_computation
```
Comment (btracey) [inline mathext/ell_complete.go]: Am I missing something? The code below looks wrong. It's using `t`, but the fortran code uses `t2`, i.e. `t*t`

Comment (btracey) [reply mathext/ell_complete.go]: The same looks true also for `CompleteE`

**Comment by reconditematter** [comment]
How interesting. I ported his Fortran code for CompleteK and CompleteE about 6 months ago and he must have reworked his code since. If we look at his original paper, on p.18 Table 7 we can see in the column K1 the same coefficients I used in https://github.com/gonum/gonum/blob/996b88e8f8941c9defdbb338bea5e239a230e742/mathext/ell_complete.go#L26 I could rewrite my code for K and E by porting his new code, maybe he decided it to be more numerically accurate. However, the current code passes all tests, and I did a spot check for several values against Mathematica (online).

**Comment by btracey** [comment]
Interesting. Maybe we can write him a note to ask. Any opinions @kortschak ? The substance of this PR LGTM

**Comment by btracey** [comment]
Ping @kortschak ?

**Comment by kortschak** [comment]
I have no strong views, but it might be worth contacting him.

**Comment by kortschak** [comment]
I'm going to merge this on the basis that there will be a follow-up PR after contacting the original author.

**Pull request merged**